File: madameautomata-bot-python-autoreg/bot/commands/release.py

import discord
from discord import app_commands, Interaction
from ..theme import gothic_embed
from ..util.perm import is_domme, is_staff, is_sub
from ..db import Session, User, Collaring, Cage, JailSession, Star, Task, Worship, Config, CollarType, StarStatus
from ..util.time import parse_duration, random_case
from ..util.ids import cuid
from datetime import datetime, timedelta
from sqlalchemy import select, func, update, and_

def setup(bot):
    @bot.tree.command(name="release", description="Release a user from collar, chastity, or jail.")
    @app_commands.describe(user="Target", what="What to release")
    @app_commands.choices(what=[
        app_commands.Choice(name="Collar", value="collar"),
        app_commands.Choice(name="Chastity", value="chastity"),
        app_commands.Choice(name="Jail", value="jail"),
        app_commands.Choice(name="Brat Jail", value="bratjail"),
    ])
    async def release(i: Interaction, user: discord.User, what: app_commands.Choice[str]):
        member = i.guild.get_member(i.user.id)
        if not await is_domme(member):
            return await i.response.send_message("Only Dommes can use this.", ephemeral=True)
        async with Session() as s:
            if what.value == "collar":
                res = await s.execute(select(Collaring).where(Collaring.sub_id==user.id, Collaring.domme_id==i.user.id, Collaring.active==True))
                active = res.scalars().first()
                if not active:
                    return await i.response.send_message("You don't have an active collar on this sub.", ephemeral=True)
                active.active = False
                active.ended_at = datetime.utcnow()
                u = await s.get(User, user.id)
                if u:
                    u.domme_releases += 1
                await s.commit()
                return await i.response.send_message(embed=gothic_embed("Collar Released", f"Released <@{user.id}>."))
            if what.value == "chastity":
                await s.execute(update(Cage).where(Cage.sub_id==user.id, Cage.domme_id==i.user.id, Cage.active==True).values(active=False))
                await s.commit()
                return await i.response.send_message(embed=gothic_embed("Chastity Released", f"Released <@{user.id}> from chastity."))
            if what.value in ("jail","bratjail"):
                await s.execute(update(JailSession).where(JailSession.sub_id==user.id, JailSession.active==True).values(active=False))
                await s.commit()
                return await i.response.send_message(embed=gothic_embed("Jail Released", f"Released <@{user.id}> from jail."))

File: madameautomata-js-bot/src/commands/release.js

import { SlashCommandBuilder, PermissionFlagsBits, ChannelType, ActionRowBuilder, ButtonBuilder, ButtonStyle, StringSelectMenuBuilder, EmbedBuilder } from "discord.js";
import { prisma } from "../lib/prisma.js";
import { Theme, gothicEmbed } from "../theme/theme.js";
import { parseDuration, randomCase } from "../util/duration.js";
import { cuid } from "../util/ids.js";
import { isDomme, isSub, isStaff, getConfig } from "../util/perm.js";

const cmd = 
{
  data: new SlashCommandBuilder()
    .setName("release")
    .setDescription("Release a user from collar, chastity, or jail.")
    .addUserOption(o=>o.setName("user").setDescription("Target").setRequired(true))
    .addStringOption(o=>o.setName("what").setDescription("What to release").setRequired(true).addChoices(
      { name: "Collar", value: "collar" },
      { name: "Chastity", value: "chastity" },
      { name: "Jail", value: "jail" },
      { name: "Brat Jail", value: "bratjail" }
    )),
  async run(i) {
    if (!await isDomme(i.member)) return i.reply({ content: "Only Dommes can use this.", ephemeral: true });
    const user = i.options.getUser("user", true);
    const what = i.options.getString("what", true);
    if (what === "collar") {
      const active = await prisma.collaring.findFirst({ where: { subId: user.id, dommeId: i.user.id, active: true } });
      if (!active) return i.reply({ content: "You don't have an active collar on this sub.", ephemeral: true });
      await prisma.collaring.update({ where: { id: active.id }, data: { active: false, endedAt: new Date() } });
      await prisma.user.update({ where: { id: user.id }, data: { dommeReleases: { increment: 1 } } }).catch(()=>{});
      return i.reply({ embeds: [gothicEmbed("Collar Released", `Released <@${user.id}>.`)] });
    }
    if (what === "chastity") {
      await prisma.cage.updateMany({ where: { subId: user.id, dommeId: i.user.id, active: true }, data: { active: false } });
      return i.reply({ embeds: [gothicEmbed("Chastity Released", `Released <@${user.id}> from chastity.`)] });
    }
    if (what === "jail" || what === "bratjail") {
      await prisma.jailSession.updateMany({ where: { subId: user.id, active: true }, data: { active: false } });
      return i.reply({ embeds: [gothicEmbed("Jail Released", `Released <@${user.id}> from jail.`)] });
    }
  }
}


export default cmd;
